### DISCUSSION TITLE
Is there a way to get rust-gpu to work with nalgebra types?

### DISCUSSION THREAD

**Discussion by marstaik** [Rust-GPU/rust-gpu#41] [General]
Hey guys, I love the idea of the project. My question is if its possible to replace the glam:: types with those from nalgebra.

I prefer to use more mathematically based libraries (and nalgebra has glm-like wrappers too from `nalgebra_glm`.

It would be amazing if I didn't have to convert to glam::Vec types to pass around data for use on cpu/gpu shaders to minimize dependencies and keep my code paths clean and easy to understand.

I don't fully understand how the glam:: types get converted to SPIR-V, I looked a bit through the code and I couldn't fully find if there is any function/translation mapping going on.

Is this currently possible? Would it be a huge lift to do?

In addition to ergonomics, there are some cpu-side performance benefits of nalgebra on certain operations that are of interest.

**Comment by schell** [comment]
I don't see anything about the types themselves that would keep them from being compiled with `rust-gpu`. Typically the first hurdle is making sure the crate can be compiled as `no_std`. After that there are certain operations to avoid, such as anything that uses `NaN` or infinities. The best way to find out is to try it yourself! If you run into any problems please open a ticket here and we'll see what we can do :)

**Comment by schell** [comment]
You can check out this quick list of gotchas that I've run into [here](https://github.com/schell/renderling/blob/main/NOTES.md#cons--limititions--gotchas). It's _mostly_ up to date.

**Reply by LegNeato** [reply]
@schell Your link does not work

**Reply by schell** [reply]
Thanks @LegNeato, I've updated the link.

**Comment by marstaik** [comment]
I've been trying to make progress on this again.. I've gotten decently far but I hit into a snag with a ptr_offset_from intrinsic:

```
   Compiling rtd_scene v0.1.0 (/home/marios/proj/rtd/packages/rtd_scene)
error: failed to run custom build command for `rtd_scene v0.1.0 (/home/marios/proj/rtd/packages/rtd_scene)`

Caused by:
  process didn't exit successfully: `/home/marios/proj/rtd/target/debug/build/rtd_scene-d05018d3cb45663e/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=RUSTGPU_CODEGEN_ARGS
  cargo:rerun-if-env-changed=RUSTGPU_RUSTFLAGS
  cargo:rerun-if-env-changed=RUSTGPU_CARGOFLAGS

  --- stderr
     Compiling rtd_scene_shaders v0.1.0 (/home/marios/proj/rtd/packages/rtd_scene_shaders)
  error: exactsdiv() not implemented for type u32
     --> /home/marios/.rustup/toolchains/nightly-2025-06-30-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/const_ptr.rs:671:18
      |
  671 |         unsafe { intrinsics::ptr_offset_from(self, origin) }
      |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  error: could not compile `rtd_scene_shaders` (lib) due to 1 previous error
  Error: BuildFailed
  ```
  
  I've managed to setup nalgebra like this in my no_std crate:
  ```
  [package]
name = "rtd_math"
version.workspace = true
edition.workspace = true


[target.'cfg(not(target_arch = "spirv"))'.dependencies]
nalgebra = { version = "0.34.1", features = ["bytemuck", "glam030", "macros"] }
nalgebra-glm = { version = "0.20.0" }

[target.'cfg(target_arch = "spirv")'.dependencies]
nalgebra = { version = "0.34.1", features = ["libm-force"], default-features = false }
nalgebra-glm = { version = "0.20.0", default-features = false }
```

Is this something we can get from spirv?

**Reply by marstaik** [reply]
Augh, even when removing the `try_inverse()` call there are many more errors, I'll have to keep trying

**Reply by Firestar99** [reply]
Currently, pointers are not super well supported, so you may be running into nalgebra using them in a way rust-gpu can't quite understand them yet. [See what I wrote here](https://github.com/Rust-GPU/rust-gpu/discussions/466#discussioncomment-15084739)
